aws-lambda/05_UpdatingOurFunction.md

```diff
@@ -0,0 +1,94 @@
+
+<!---[CodeSnippet](step4HttpMaidConfig)-->
+```java
+private static QuantumMaid quantumMaidConfig() {
+  final QuantumMaid quantumMaid = QuantumMaid.quantumMaid()
+      .get("/hello/<whoever-you-are>", (request, response) -> response.setBody(
+          String.format("Hello %s!",
+              request.pathParameters().getPathParameter("whoever-you-are"))));
+  return quantumMaid;
+}
+```
+
+## Updating the function using SAM CLI
+
+To deploy the change, use the same command pair as in step 3.
+This time however, we don't need to answer any questions, because previous answers are saved to `samconfig.toml`
+
+```shell
+$ sam build
+$ sam deploy
+...
+Successfully created/updated stack - hello-app in us-east-1
+```
+
+## Viewing logs using SAM CLI
+
+Before we invoke the URL, let's tail the logs, which will give us some interesting details on our function execution(s).
+In a separate terminal, run:
+
+```shell
+$ sam logs --tail --region us-east-1 --stack-name hello-app --name HelloWorldFunction
+```
+
+## Getting runtime statistics
+
+Back in the terminal where you ran `sam deploy`, run:
+
+```shell
+$ time echo $(curl -s $apiUrl/hello/{me,you,everyone})
+Hello me!Hello you!Hello everyone!
+
+real	0m2.990s
+user	0m0.077s
+sys	0m0.005s
+```
+
+The logs display something similar to the following simplified output:
+
+```text
+... 2020-06-21T09:22:07.659000 START RequestId: ... Version: $LATEST
+... 2020-06-21T09:22:07.677000 {version=2.0, routeKey=ANY /{proxy+}, rawPath=/hello/me,...
+... 2020-06-21T09:22:08.058000 END RequestId: ...
+... 2020-06-21T09:22:08.058000 REPORT RequestId: ...	Duration: 398.08 ms	Billed Duration: 400 ms	Memory Size: 512 MB	Max Memory Used: 101 MB	Init Duration: 870.58 ms
+... 2020-06-21T09:22:08.242000 START RequestId: ... Version: $LATEST
+... 2020-06-21T09:22:08.246000 {version=2.0, routeKey=ANY /{proxy+}, rawPath=/hello/you,...
+... 2020-06-21T09:22:08.248000 END RequestId: ...
+... 2020-06-21T09:22:08.248000 REPORT RequestId: ...	Duration: 3.32 ms	Billed Duration: 100 ms	Memory Size: 512 MB	Max Memory Used: 101 MB
+... 2020-06-21T09:22:08.442000 START RequestId: ... Version: $LATEST
+... 2020-06-21T09:22:08.446000 {version=2.0, routeKey=ANY /{proxy+}, rawPath=/hello/everyone,...
+... 2020-06-21T09:22:08.448000 END RequestId: ...
+... 2020-06-21T09:22:08.448000 REPORT RequestId: ...	Duration: 3.32 ms	Billed Duration: 100 ms	Memory Size: 512 MB	Max Memory Used: 101 MB
+```
+
+Important statistics:
+
+- Time spent in Lambda = Init Duration: 870.58 ms + Duration: 398.08 ms + Duration: 3.32 ms + Duration: 3.32 ms = 1275.3 ms
+- Time spent overall = 2.990 s = 2990 ms
+- Unaccounted time = 2990 ms - 1275.3 ms = 1714.7 ms, of which
+  - approx. 600 ms is probably due to network latency (100 ms each way x 3 requests)
+  - approx. 1100 ms is as yet unexplained
+
+Having this log output at hand allows us to draw a few conclusions:
+
+- The HttpMaid overhead is flat beyond the first invocation (<5 ms)
```

Comment:
HttpMaid -> QuantumMaid
This also the only item in this list that does not end with a period